### PR TITLE
Add "TimeToSign" performance tracking

### PR DESCRIPTION
**Pull request by walmat**
Fixes APP-1705

## What changed (plus any additional context for devs)
Most of the details for this ticket are laid out in the [linear ticket](https://linear.app/rainbow/issue/APP-1705/track-metrics-for-time-to-sign-a-transaction). This adds time tracking to the following flows:
- swap from tap to swap -> sheet dismissal
- send from confirm CTA -> sheet dismissal
- WC signs / txns from CTA -> sheet dismissal
- Dapp Browser txns from CTA -> sheet dismissal

Important to note that the performanceTracking store works with worklet functions and js functions. Nothing extra needs configured, just pass the function as the `fn` prop.

## Screen recordings / screenshots

<details>
  <summary>Expand Logs</summary>

  ### Dapp Browser
```
 LOG  12:52:19 [DEBUG] [performance]: Time to complete operation {
  "log": {
    "completedAt": 1722444739175,
    "screen": "DappBrowser",
    "operation": "CallToAction",
    "startTime": 6621080562.915833,
    "endTime": 6621080635.8495,
    "timeToCompletion": 72.93366622924805
  }
   LOG  12:52:20 [DEBUG] [performance]: Time to complete operation {
  "log": {
    "completedAt": 1722444740156,
    "screen": "DappBrowser",
    "operation": "Authentication",
    "startTime": 6621081615.058333,
    "endTime": 6621081617.169375,
    "timeToCompletion": 2.111042022705078
  }
}
 LOG  12:52:20 [DEBUG] [performance]: Time to complete operation {
  "log": {
    "completedAt": 1722444740221,
    "screen": "DappBrowser",
    "operation": "KeychainRead",
    "startTime": 6621081614.936416,
    "endTime": 6621081682.671458,
    "timeToCompletion": 67.73504257202148
  }
}
 LOG  12:52:22 [DEBUG] [performance]: Time to complete operation {
  "log": {
    "completedAt": 1722444742290,
    "screen": "DappBrowser",
    "operation": "BroadcastTransaction",
    "startTime": 6621081683.86225,
    "endTime": 6621083751.078375,
    "timeToCompletion": 2067.216124534607
  }
}
 LOG  12:52:24 [DEBUG] [performance]: Time to complete operation {
  "log": {
    "completedAt": 1722444744335,
    "screen": "DappBrowser",
    "operation": "SheetDismissal",
    "startTime": 6621085715.521458,
    "endTime": 6621085787.523375,
    "timeToCompletion": 72.00191688537598
  }
}
 LOG  12:52:24 [DEBUG] [performance]: Time to sign {
  "screen": "DappBrowser",
  "elapsedTime": 2281.9977922439575,
  "completedAt": 1722444744336
}
 ```
</details>

 

## What to test
in order to test, please set **LOG_LEVEL** to **debug**
- [x] confirm that swaps has no regressions and is tracked
- [x] confirm that sends has no regressions and is tracked
- [x] confirm that walletconnect has no regressions and is tracked
- [x] confirm that dapp browser has no regressions and is tracked

### CODE REVIEW

File: src/model/wallet.ts [hunk 1]
```diff
@@ -257,11 +258,17 @@ export const walletInit = async (
   return { isNew, walletAddress };
 };
 
-export const loadWallet = async (
-  address?: EthereumAddress | undefined,
-  showErrorIfNotLoaded = true,
-  provider?: Provider
-): Promise<null | Wallet | LedgerSigner> => {
+export const loadWallet = async <S extends Screen>({
```
Comment (walmat) [inline src/model/wallet.ts:261]: The following is really just converting the `loadWallet` method to object parameters so that we can pass more than one optional.

File: src/hooks/useENSRegistrationActionHandler.ts [hunk 2]
```diff
@@ -313,7 +336,7 @@ export default function useENSRegistrationActionHandler(
   );
 
   return {
-    action: actions[registrationStep] as (...args: any) => void,
+    action: actions[registrationStep],
```
Comment (walmat) [inline src/hooks/useENSRegistrationActionHandler.ts:339]: removed this to stop from clobbering the actual return types of the functions.

File: src/hooks/useENSRegistrationActionHandler.ts
Comment (walmat) [inline src/hooks/useENSRegistrationActionHandler.ts:1]: Gist of these file changes are just from converting `loadWallet` to obj params

File: src/raps/execute.ts [hunk 3]
```diff
@@ -125,50 +130,45 @@ export const walletExecuteRap = async (
...
-      index: 0,
-      baseNonce: nonce,
-      rapName,
-      flashbots: parameters?.flashbots,
-      gasParams: parameters?.gasParams,
-      gasFeeParamsBySpeed: parameters?.gasFeeParamsBySpeed,
-    };
 
-    const { baseNonce, errorMessage: error, hash } = await executeAction(actionParams);
+  if (actions.length) {
```
Comment (walmat) [inline src/raps/execute.ts:148]: To be honest I'm not sure why we extract the first action out and then loop over the rest. Happy to revert this change, but it didn't make any sense to me why we had this here.

Also, fwiw I tested with swap & swap+unlock and both worked just fine.

File: src/model/wallet.ts [hunk 4]
```diff
@@ -257,11 +258,17 @@ export const walletInit = async (
...
+  address,
+  showErrorIfNotLoaded = false,
```
Comment (jinchung) [inline src/model/wallet.ts:263]: checking on why the change in default from `true` to `false` for this param

Comment (walmat) [reply src/model/wallet.ts:263]: Whoops good catch! Let me fix that and the default instances

Comment (walmat) [reply src/model/wallet.ts]: Fixed: https://github.com/rainbow-me/rainbow/pull/5962/commits/a0dbfc61d6f4bd560da1b06c57d14337c8258718

File: src/raps/execute.ts (hunk 3 above)
Comment (walmat) [reply src/raps/execute.ts]: Reverted, this can be part of a follow-up PR and doesn't need to be in this PR.

File: src/model/wallet.ts [hunk 5]
```diff
@@ -257,11 +258,17 @@ export const walletInit = async (
...
-  address?: EthereumAddress | undefined,
+export const loadWallet = async <S extends Screen>({
+  address,
   showErrorIfNotLoaded = true,
-  provider?: Provider
-): Promise<null | Wallet | LedgerSigner> => {
+  provider = web3Provider,
+  timeTracking,
+}: {
+  address: EthereumAddress | undefined;
```
Comment (jinchung) [inline src/model/wallet.ts:267]: any reason not to keep the `?` optional on `address` so you don't have to pass the `address` in as undefined explicitly everywhere?

File: src/model/wallet.ts [hunk 6]
```diff
@@ -257,11 +258,17 @@ export const walletInit = async (
...
 };
 
-export const loadWallet = async (
-  address?: EthereumAddress | undefined,
+export const loadWallet = async <S extends Screen>({
+  address,
   showErrorIfNotLoaded = true,
-  provider?: Provider
-): Promise<null | Wallet | LedgerSigner> => {
+  provider = web3Provider,
```
Comment (jinchung) [inline src/model/wallet.ts:264]: I know this is likely a noop but let's leave these default changes out of this PR for now to reduce scope for testing for QA, and any improvements we want to make we can make a separate PR for (e.g, keeping `provider` as optional and not defaulting it to the `web3Provider`)

File: src/screens/points/claim-flow/ClaimRewardsPanel.tsx [hunk 7]
```diff
@@ -262,7 +262,10 @@ const ClaimingRewards = ({
       } satisfies RapSwapActionParameters<'claimBridge'>;
 
       const provider = getProviderForNetwork(Network.optimism);
-      const wallet = await loadWallet(address, false, provider);
+      const wallet = await loadWallet({
+        address,
+        provider,
```
Comment (jinchung) [inline src/screens/points/claim-flow/ClaimRewardsPanel.tsx:267]: given the defaults going back to `true`, you will need to explicitly provide `showErrorIfNotLoaded` as `false` here

File: src/model/wallet.ts (hunk 6 above)
Comment (walmat) [reply src/model/wallet.ts:264]: It already defaults to web3Provider if you don't pass it later in this code block. This was mostly just to clean up typescript types.

File: src/model/wallet.ts (hunk 5 above)
Comment (walmat) [reply src/model/wallet.ts:267]: Yeah was going to add that, let me do it

File: src/screens/ExchangeModal.tsx [hunk 8]
```diff
@@ -402,7 +402,10 @@ export default function ExchangeModal({ fromDiscover, ignoreInitialTypeCheck, te
       try {
         // load the correct network provider for the wallet
         const provider = getProviderForNetwork(currentNetwork);
-        let wallet = await loadWallet(accountAddress, false, provider);
+        let wallet = await loadWallet({
+          showErrorIfNotLoaded: false,
```
Comment (jinchung) [inline src/screens/ExchangeModal.tsx:406]: this one is missing the `accountAddress` being passed along

File: src/model/wallet.ts (hunk 6 above)
Comment (jinchung) [reply src/model/wallet.ts:264]: yeah understood, it's likely a noop but it just technically introduces a potential path change difference in the event of a hardware wallet call that isn't passing the `provider` (on line 298 , which shouldn't happen) - basically what I'm getting at directionally is: since this is a sensitive PR in terms of touching core txn flows and the wallet model, let's reduce all variables as much as possible except for the changes strictly needed, and we can come back and improve the rest in a separate test

Comment (walmat) [reply src/model/wallet.ts]: 👍🏼 reverted

File: src/screens/SignTransactionSheet.tsx [hunk 9]
```diff
@@ -505,17 +523,37 @@ export const SignTransactionSheet = () => {
       return;
     }
 
-    const existingWallet = await loadWallet(accountInfo.address, true, provider);
+    const existingWallet = await performanceTracking.getState().executeFn({
+      fn: loadWallet,
```
Comment (jinchung) [inline src/screens/SignTransactionSheet.tsx:527]: wondering if there is a collision here as we'll be logging the "outer" `loadWallet` function as well as the "inner" `loadPrivateKey` function inside of `loadWallet` with the same screen / operation name that we're passing into `timeTracking` for `loadWallet`.

File: src/screens/SignTransactionSheet.tsx [hunk 10]
```diff
@@ -765,7 +822,7 @@ export const SignTransactionSheet = () => {
     if (accountInfo.isHardwareWallet) {
       navigate(Routes.HARDWARE_WALLET_TX_NAVIGATOR, { submit: onPressSend });
     } else {
-      await onPressSend();
+      onPressSend();
```
Comment (jinchung) [inline src/screens/SignTransactionSheet.tsx:825]: checking if you wanted to remove this `await` - would it be more fair to make sure that the performance tracking for `submitFn` includes the completion of `onPressSend` ?

File: src/screens/SignTransactionSheet.tsx (hunk 9 above)
Comment (walmat) [reply src/screens/SignTransactionSheet.tsx:527]: Good catch! Should be Authentication for the inner one.

File: src/screens/SignTransactionSheet.tsx (hunk 10 above)
Comment (walmat) [reply src/screens/SignTransactionSheet.tsx:825]: I wanted to separate the actual time to call the CTA instead of waiting for the inner functions to complete. Technically I think this is more of what we want?

If we await, we'll be awaiting the entire txn process.

File: src/screens/SignTransactionSheet.tsx [hunk 11]
```diff
@@ -946,7 +1003,9 @@ export const SignTransactionSheet = () => {
                         : i18n.t(i18n.l.walletconnect.simulation.buttons.confirm)
                     }
                     newShadows
-                    onPress={submitFn}
+                    onPress={performanceTracking
```
Comment (jinchung) [inline src/screens/SignTransactionSheet.tsx:1006]: in a similar vein to my comment above - could we put the `performanceTracking....executeFn` inside of the `submitFn` callback

File: src/screens/SignTransactionSheet.tsx (hunk 10 above)
Comment (walmat) [reply src/screens/SignTransactionSheet.tsx:825]: We could do that, just need to add other operations to track that "other stuff" part

File: src/screens/SendSheet.js [hunk 12]
```diff
@@ -427,7 +439,7 @@ export default function SendSheet(props) {
       let nextNonce;
 
       if (isENS && toAddress && (clearRecords || setAddress || transferControl)) {
-        const { nonce } = await transferENS(() => null, {
+        const { nonce } = await transferENS({
```
Comment (jinchung) [inline src/screens/SendSheet.js:442]: 😅

File: src/screens/SendSheet.js [hunk 13]
```diff
@@ -458,7 +470,11 @@ export default function SendSheet(props) {
       };
 
       try {
-        const signableTransaction = await createSignableTransaction(txDetails);
+        const signableTransaction = await performanceTracking.getState().executeFn({
+          fn: createSignableTransaction,
+          operation: TimeToSignOperation.SignTransaction,
```
Comment (jinchung) [inline src/screens/SendSheet.js:475]: just a heads up that the `createSignableTransaction` doesn't sign the transaction itself, it basically just creates the txn payload data - so this will be a bit discrepant from the other places we use `TimeToSignOperation.SignTransaction` which actually load the wallet and sign the txn. The `sendTransaction` below is where the signing and broadcasting will happen, so we may be covered just from that.

File: src/screens/SignTransactionSheet.tsx [hunk 14]
```diff
@@ -454,10 +460,20 @@ export const SignTransactionSheet = () => {
...
+    nativeCurrency,
+    req,
+    request.message,
+    simulationUnavailable,
+    transactionDetails,
+  ]);
 
-  const closeScreen = useCallback(
-    (canceled: boolean) => {
+  const closeScreen = performanceTracking.getState().executeFn({
```
[the remark this answers is not in the document]
Comment (walmat) [reply src/screens/SignTransactionSheet.tsx:475]: it needs to wrap a function, so I'd have to create another function inside of the useCallback in order to wrap it? Not sure if that's preferred.

File: src/screens/SignTransactionSheet.tsx (hunk 11 above)
Comment (walmat) [reply src/screens/SignTransactionSheet.tsx:1006]: > it needs to wrap a function, so I'd have to create another function inside of the useCallback in order to wrap it? Not sure if that's preferred.

File: src/screens/SendConfirmationSheet.tsx [hunk 15]
```diff
@@ -376,25 +377,29 @@ export const SendConfirmationSheet = () => {
...
-      if (isENS) {
-        const clearRecords = checkboxes.some(({ checked, id }) => checked && id === 'clear-records');
-        const setAddress = checkboxes.some(({ checked, id }) => checked && id === 'set-address');
-        const transferControl = checkboxes.some(({ checked, id }) => checked && id === 'transfer-control');
-        await callback({
-          ens: { clearRecords, setAddress, transferControl },
-        });
-      } else {
-        await callback();
+  const handleSubmit = performanceTracking.getState().executeFn({
```
Comment (jinchung) [inline src/screens/SendConfirmationSheet.tsx:380]: same thing here for keeping `useCallback`

File: src/raps/actions/ens.ts [hunk 16]
```diff
@@ -673,7 +675,11 @@ const executeAction = async (
   try {
     logger.log('[2 INNER] executing type', type);
     const actionPromise = findENSActionByType(type);
-    nonce = await actionPromise(wallet, rap, index, parameters as RapENSActionParameters, baseNonce);
+    nonce = await performanceTracking.getState().executeFn({
+      fn: actionPromise,
```
Comment (jinchung) [inline src/raps/actions/ens.ts:679]: this will take a different path during an error: the actions from `findENSActionByType` can throw errors, which will get swallowed by the `executeFn` and return `undefined` instead of throwing and ending in the catch

File: src/raps/actions/ens.ts [hunk 17]
```diff
@@ -673,7 +675,11 @@ const executeAction = async (
...
+      screen: Routes.SEND_SHEET,
```
Comment (jinchung) [inline src/raps/actions/ens.ts:680]: did you want a different screen for this ENS action?

File: src/raps/actions/ens.ts [hunk 18]
```diff
@@ -700,7 +706,11 @@ export const executeENSRap = async (
   parameters: ENSActionParameters,
   callback: (success?: boolean, errorMessage?: string | null) => void
 ) => {
-  const rap = await createENSRapByType(type, parameters as ENSActionParameters);
+  const rap = await performanceTracking.getState().executeFn({
+    fn: createENSRapByType,
+    operation: TimeToSignOperation.CreateRap,
+    screen: Routes.SEND_SHEET,
```
Comment (jinchung) [inline src/raps/actions/ens.ts:712]: same here: did you want a different screen for this ENS action?

File: src/raps/execute.ts [hunk 19]
```diff
@@ -125,7 +131,18 @@ export const walletExecuteRap = async (
   type: RapTypes,
   parameters: RapSwapActionParameters<'swap' | 'crosschainSwap' | 'claimBridge'>
 ): Promise<{ nonce: number | undefined; errorMessage: string | null }> => {
-  const rap: Rap = await createSwapRapByType(type, parameters);
+  // NOTE: We don't care to track claimBridge raps
+  const rap =
```
Comment (jinchung) [inline src/raps/execute.ts:135]: `createSwapByRap` type can also throw errors, so these will be swallowed up and will throw a different error when it tries to read `{ actions }` from `undefined`

File: src/screens/SendConfirmationSheet.tsx [hunk 20]
```diff
@@ -376,25 +377,33 @@ export const SendConfirmationSheet = () => {
...
+      performanceTracking.getState().executeFn({
+        fn: async () => {
+          if (!canSubmit) return;
+          try {
+            setIsAuthorizing(true);
+            if (isENS) {
+              const clearRecords = checkboxes.some(({ checked, id }) => checked && id === 'clear-records');
+              const setAddress = checkboxes.some(({ checked, id }) => checked && id === 'set-address');
+              const transferControl = checkboxes.some(({ checked, id }) => checked && id === 'transfer-control');
+              callback({
```
Comment (jinchung) [inline src/screens/SendConfirmationSheet.tsx:391]: can we bring back the `awaits` here so that we don't change the flow of handling via the try/catch

File: src/hooks/useENSRegistrationActionHandler.ts [hunk 21]
```diff
@@ -120,21 +159,20 @@ export default function useENSRegistrationActionHandler(
...
+  const speedUpCommitAction: ActionTypes[typeof REGISTRATION_STEPS.WAIT_COMMIT_CONFIRMATION] = useCallback(
     async (accentColor: string) => {
-      // we want to speed up the last commit tx sent
       const commitTransactionHash = registrationParameters?.commitTransactionHash;
 
       const tx = getPendingTransactionByHash(commitTransactionHash || '');
       commitTransactionHash &&
         tx &&
-        navigate(ios ? Routes.SPEED_UP_AND_CANCEL_SHEET : Routes.SPEED_UP_AND_CANCEL_BOTTOM_SHEET, {
+        navigate(Routes.SPEED_UP_AND_CANCEL_SHEET, {
```
Comment (jinchung) [inline src/hooks/useENSRegistrationActionHandler.ts:175]: confirming why the removal of `ios` check here

File: src/hooks/useENSRegistrationActionHandler.ts [hunk 22]
```diff
@@ -291,31 +341,32 @@ export default function useENSRegistrationActionHandler(
         transferControl,
       };
 
-      const { nonce: newNonce } = await executeENSRap(wallet, ENSRapActionType.transferENS, transferEnsParameters, callback);
+      const { nonce: newNonce } = await performanceTracking.getState().executeFn({
+        fn: executeENSRap,
+        screen: Routes.SEND_SHEET,
```
Comment (jinchung) [inline src/hooks/useENSRegistrationActionHandler.ts:346]: checking if you wanted a different `screen` here

**Review by jinchung — APPROVED** [review]
🌮
